source.go

```diff
@@ -1,60 +1,146 @@
+	client  *http.Client
+	url     string
+	limiter *rate.Limiter
 }
 
 func NewSource() sdk.Source {
 	return &Source{}
 }
 
 func (s *Source) Parameters() map[string]sdk.Parameter {
-	return map[string]sdk.Parameter{
-		GlobalConfigParam: {
-			Default:     "localhost:10000",
-			Required:    true,
-			Description: "The URL of the server.",
-		},
-		SourceConfigParam: {
-			Default:     "false",
-			Required:    false,
-			Description: "An optional configuration parameter.",
-		},
-	}
+	return s.config.Parameters()
 }
 
 func (s *Source) Configure(ctx context.Context, cfg map[string]string) error {
-	sdk.Logger(ctx).Info().Msg("Configuring a Source Connector...")
-	config, err := ParseSourceConfig(cfg)
+	sdk.Logger(ctx).Info().Msg("Configuring Weather Source Connector...")
+	var config SourceConfig
+	err := sdk.Util.ParseConfig(cfg, &config)
 	if err != nil {
 		return err
 	}
-
 	s.config = config
 	return nil
 }
 
 func (s *Source) Open(ctx context.Context, pos sdk.Position) error {
+	s.client = &http.Client{}
+	s.url = s.CreateRequestURL()
+	// try pinging the URL with APPID
+	req, err := http.NewRequestWithContext(ctx, http.MethodHead, s.url, nil)
+	if err != nil {
+		return fmt.Errorf("error creating HTTP request %q: %w", s.config.URL, err)
+	}
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("error pinging URL %q: %w", s.config.URL, err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == http.StatusUnauthorized {
+		return fmt.Errorf("authorization failed, check your APPID key")
+	}
+
+	s.limiter = rate.NewLimiter(rate.Every(s.config.PollingPeriod), 1)
+
 	return nil
 }
 
 func (s *Source) Read(ctx context.Context) (sdk.Record, error) {
-	return sdk.Record{}, nil
+	err := s.limiter.Wait(ctx)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("context was cancelled: %w", err)
+	}
+	rec, err := s.getRecord(ctx)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("error getting the weather data: %w", err)
+	}
+	return rec, nil
 }
 
 func (s *Source) Ack(ctx context.Context, position sdk.Position) error {
-	return nil
+	sdk.Logger(ctx).Debug().Str("position", string(position)).Msg("got ack")
+	return nil // no ack needed
 }
 
 func (s *Source) Teardown(ctx context.Context) error {
+	if s.client != nil {
+		s.client.CloseIdleConnections()
+	}
 	return nil
 }
+
+func (s *Source) CreateRequestURL() string {
+	return s.config.URL + "?" + "q=" + s.config.City + "&" + "APPID=" + s.config.APPID + "&" + "units=" + s.config.Units
+}
+
+func (s *Source) getRecord(ctx context.Context) (sdk.Record, error) {
+	// create GET request
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.url, nil)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("error creating HTTP request: %w", err)
+	}
+	// get response
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("error getting data from URL: %w", err)
+	}
+	defer resp.Body.Close()
+	// read body
+	body, err := io.ReadAll(resp.Body)
```

Comment:
We don't check the status code before reading the body, I'm not how errors look like that are returned from the API but if the error is a valid JSON string it could happen that we create a record from an error. Or at the very least we would notice it in line 132 when the error won't contain the field `dt` and the connector will produce a strange error message.

Long story short - I'd make sure that we only proceed if the status code is expected (probably 200, right?).